feat: common 모듈 공통 데이터 생성 및 util 생성

dowon0113 opened the pull request:
## 개요
공통 모듈(Global Module)에 대한 예외 처리 및 페이징/정렬 기능 생성
- 공통 예외 처리(GlobalExceptionHandler), 페이징 및 정렬(JpaUtil, QueryDSLConfig, QueryDslUtil) 기능 추가, 날짜 및 JSON 변환 기능 테스트

<br>

## 📝 작업 내용
### 변경 사항
예외 관리를 ExceptionCode Enum을 통해 일관되게 관리하도록 구조를 변경하였습니다.

1. 글로벌 예외 처리 기능 개선
- ExceptionCode Enum을 활용한 예외 코드 관리
향후 예외 코드 추가 및 확장 시 Enum만 수정하면 되어 유지보수성이 향상됨
- GlobalExceptionHandler 예외 처리 로직 개선
BusinessLogicException, RuntimeException, AuthenticationException 등 예외 로직 추가
- Spring ResponseEntityExceptionHandler를 활용한 유효성 검증 예외 처리
MethodArgumentNotValidException: 유효성 검증 실패 시, 필드별 오류 메시지를 포함하도록 개선
HttpMessageNotReadableException: 잘못된 JSON 요청을 처리하도록 개선
HandlerMethodValidationException: 메서드 레벨 검증(@Validated) 오류 처리 추가

3. 페이징 및 정렬 기능 추가
- JpaUtil : Spring Data JPA 기반으로 페이징 및 정렬 처리 기능 추가
    - 요청된 페이지 번호, 페이지 크기, 정렬 기준을 기반으로 PageRequest 객체 생성
    - CREATED, UPDATED 기준으로만 정렬 가능하도록 제한
- QueryDSLConfig : JPAQueryFactory 빈 등록하여 QueryDSL 사용 가능하도록 설정
- QueryDslUtil : QueryDSL 기반으로 동적 정렬 기능 추가
    - Pageable 정보를 기반으로 다중 정렬 지원
    - OrderSpecifier<?> 를 활용하여 QueryDSL 정렬 조건 생성

4. 공통 유틸리티 클래스 추가
- CommonUtil : 날짜 변환 및 JSON → DTO 변환 기능 추가
- CommonUtilTest : 날짜 변환 기능 테스트, 문자열 빈 값 확인, JSON → DTO 변환 기능 테스트 추가

5. QueryDSL 자동 생성 파일 Git 제외 처리
QBaseDataEntity.java 같은 자동 생성 파일을 .gitignore에 추가하여 Git에서 추적하지 않도록 변경


### 추가 설명
다음부턴 커밋 더 쪼개볼게요! ㅠㅠ
<br>

### 💬리뷰 요구사항

<br>

### #️⃣ 연관된 이슈
closed #23

<br>

## PR 유형

이 PR은 어떤 변경 사항을 포함하고 있나요?

- [X] 새로운 기능 추가
- [ ] 버그 수정
- [ ] UI 디자인 변경 (CSS 등)
- [ ] 코드에 영향을 주지 않는 변경사항 (오타 수정, 탭 사이즈 변경, 변수명 변경 등)
- [X] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [X] 문서 수정
- [X] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 또는 폴더명 수정
- [ ] 파일 또는 폴더 삭제
- [ ] 배포 준비, PR 관련 사항

<br>

## ✅ Check List

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [X] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [X] 변경 사항에 대한 테스트를 했습니다. (버그 수정/기능에 대한 테스트)
- [ ] CI/CD 파이프라인을 통과했습니다.

dowon0113 committed:
feat: CommonDto, BaseDataEntity 생성

- 모든 엔티티에서 사용할 공통 필드 추가

Resolved #23

dowon0113 committed:
feat: CommonUtil 생성

CommonUtil 추가 (JSON을 DTO 변환, 날짜 변환 기능 포함)

Resolved #23

dowon0113 committed:
Merge branch 'develop' of https://github.com/3babko5/babko into feat/common

dowon0113 committed:
refactor: ID 타입을 UUID로 변경

ID 타입을 UUID로 변경하였음

Resolved #23

dowon0113 committed:
feat: 글로벌 예외처리 생성

- BusinessLogicException 생성: 비즈니스 로직 관련 예외 처리를 위한 커스텀 예외 클래스 추가
- GlobalExceptionHandler 추가: 공통 예외 처리를 위한 글로벌 예외 핸들러 구현
  - 비즈니스 로직 예외 (BusinessLogicException) 처리
  - 유효성 검사 예외 (MethodArgumentNotValidException) 처리
  - 인증 예외 (AuthenticationException) 처리
  - FeignClient 예외 처리
  - 데이터베이스 관련 예외 처리
  - 서버 내부 오류 처리
- GlobalExceptionHandlerTest 추가: 주요 예외 상황에 대한 테스트 작성

Resolved #23

dowon0113 committed:
feat: 페이징 및 정렬 기능 추가 (JpaUtil, QueryDSLConfig, QueryDslUtil)

- JpaUtil : Spring Data JPA 기반으로 페이징 및 정렬 처리 기능 추가
요청된 페이지 번호, 페이지 크기, 정렬 기준을 기반으로 PageRequest 객체 생성
CREATED, UPDATED 기준으로만 정렬 가능하도록 제한
- QueryDSLConfig : JPAQueryFactory 빈 등록하여 QueryDSL 사용 가능하도록 설정
- QueryDslUtil : QueryDSL 기반으로 동적 정렬 기능 추가
Pageable 정보를 기반으로 다중 정렬 지원
OrderSpecifier<?> 를 활용하여 QueryDSL 정렬 조건 생성

Resolved #23

dowon0113 committed:
feat: CommonUtilTest 생성

- 날짜 변환 기능 테스트 (LDTToString, stringToLDT)
- 문자열 빈 값 확인 테스트 (checkStringIsEmpty): null 또는 "" (빈 문자열) 입력 시 true 반환 확인
- JSON을 DTO로 변환 기능 테스트 (convertJsonToDto)

Resolved #23

dowon0113 committed:
chore: QueryDSL 생성 파일 .gitignore에 추가

- QBaseDataEntity.java 같은 Q클래스는 빌드 시 자동 생성되므로 직접 커밋하면 불필요한 충돌이 발생하여 .gitignore에 추가

Resolved #23

Jungdaye89 commented:
공통 모듈 생성해 주시느라 고생 많이하셨습니다!

지금 GlobalExceptionHandler에서 에러 메시지를 직접 하드코딩하고 있는데, Enum으로 따로 관리하면 더 깔끔하고 유지보수하기 쉬울 것 같습니다!

AnJungGeon commented:
팀장님 수고 많으셨습니다!

에러 메세지는 Enum으로 관리하면 Swagger로 문서화 하기 용이하다고 하네요. 이 부분 고려해보면 좋을거 같습니다!

dowon0113 committed:
feat: 비즈니스 로직 예외 클래스, ExceptionCode 생성

- 공통 비즈니스 로직 예외(`BusinessLogicException`) 정의
- ExceptionCode를 활용하여 예외 코드 및 메시지를 통합 관리

Resolved #23

dowon0113 committed:
feat: 글로벌 예외 코드(Enum) 추가

- 공통 예외 코드를 관리하는 GlobalExceptionCode를 Enum으로 정의

Resolved #23

dowon0113 committed:
feat: 예외 처리 기능 추가 및 개선

- BusinessLogicException, RuntimeException 등 예외 처리 추가
- MethodArgumentNotValidException, HttpMessageNotReadableException 핸들링 개선
- ExceptionResponse에 timestamp 및 details 필드 추가
- 로깅 개선 및 예외 코드 통일

Resolved #23

dowon0113 committed:
test: GlobalExceptionHandler 통합 테스트

- GlobalExceptionHandlerTest 클래스 작성하여 예외 처리 핸들러 검증
- BusinessLogicException 처리 검증
  - 상태 코드, 에러 코드, 메시지 등 응답 구조 검증
  - MethodArgumentNotValid 예외 처리 검증
  - 유효성 검증 실패 시 에러 상세 정보 포함 여부 검증
- HttpMessageNotReadable 예외 처리 검증
  - 잘못된 JSON 요청에 대한 처리 검증
- MockExceptionCode 추가
  - 테스트용 예외 코드 enum 구현

Resolved #23

dowon0113 committed:
build: 예외 처리 테스트를 위한 의존성 추가

- spring-boot-starter-validation 의존성 추가

Resolved #23

AnJungGeon commented:
확인했습니다
정말 수고 많으셨습니다!👍

Jungdaye89 commented on common/build.gradle:
```diff
@@ -21,6 +21,9 @@ dependencies {
 	testImplementation 'org.mockito:mockito-junit-jupiter'
+
+	//validation
+	implementation 'org.springframework.boot:spring-boot-starter-validation'
```
validation은 다른 모듈에서도 직접적으로 사용 가능해야 하므로 implementation이 아니라 api 로 사용해야 합니다! issue troubleshooting 보시면 제가 작성해 놓은 것 있습니다 한번 확인해 보시고 수정 부탁드립니다 :)
